Automattic/sensei
# Split scheduler to seperate classes for easier testing

This PR looks big, but it is a lot of testing.

#### Changes proposed in this Pull Request:

* In testing https://github.com/Automattic/sensei-wc-paid-courses/pull/452 I discovered an issue with how I am using Action Scheduler's `as_next_scheduled_action()`. I've fixed it so it properly reschedules when there is currently an action `in-progress`.
* I took this opportunity (prior to beta) to do the separation of the different scheduling methods (WP Cron vs Action Scheduler). Most of this PR is actually tests for the new classes. 

#### Testing instructions:

With WooCommerce installed/activated...
* Ensure you have more than 20 learner users.
* Reset the site enrolment salt option (`sensei_course_enrolment_site_salt`).
* Load a WP admin page.
* Make sure the action is scheduled in Tools > Scheduled Actions. 
* Make sure it is _rescheduled_ for each batch.

#### New Filters
- `sensei_background_job_actions` - Get a list of background job actions that are handled by the class.
- `sensei_scheduler_class` - Override the default class that implements `Sensei_Scheduler_Interface`.
